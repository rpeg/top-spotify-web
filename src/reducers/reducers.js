import { combineReducers } from 'redux';
import {
  SET_USER,
  SET_TIME_RANGE,
  TimeRanges,
  REQUEST_ARTISTS,
  RECEIVE_ARTISTS,
  REQUEST_TRACKS,
  RECEIVE_TRACKS,
  REQUEST_FEATURES,
  RECEIVE_FEATURES,
  SET_ARTIST_COUNT,
  SET_TRACK_COUNT,
  SET_GENRE_COUNT,
} from '../constants/constants';

const { LONG } = TimeRanges;

function user(state = {}, action) {
  switch (action.type) {
    case SET_USER:
      return { ...state, user: action.user };
    default:
      return state;
  }
}

function timeRange(state = LONG, action) {
  switch (action.type) {
    case SET_TIME_RANGE:
      return { ...state, timeRange: action.timeRange };
    default:
      return state;
  }
}

function artistCount(state = 10, action) {
  switch (action.type) {
    case SET_ARTIST_COUNT:
      return { ...state, artistCount: action.artistCount };
    default:
      return state;
  }
}

function trackCount(state = 15, action) {
  switch (action.type) {
    case SET_TRACK_COUNT:
      return { ...state, trackCount: action.trackCount };
    default:
      return state;
  }
}

function genreCount(state = 40, action) {
  switch (action.type) {
    case SET_GENRE_COUNT:
      return { ...state, genreCount: action.genreCount };
    default:
      return state;
  }
}

function artists(state = {}, action) {
  switch (action.type) {
    case REQUEST_ARTISTS:
      return { ...state, isFetching: true };
    case RECEIVE_ARTISTS:
      return { ...state, isFetching: false, items: action.items };
    default:
      return state;
  }
}

function artistsByTimeRange(state = {}, action) {
  switch (action.type) {
    case RECEIVE_ARTISTS:
    case REQUEST_ARTISTS:
      return { ...state, [action.timeRange]: artists(state[action.timeRange], action) };
    default:
      return state;
  }
}

function tracks(state = {}, action) {
  switch (action.type) {
    case REQUEST_TRACKS:
      return { ...state, isFetching: true };
    case RECEIVE_TRACKS:
      return { ...state, isFetching: false, items: action.tracks };
    default:
      return state;
  }
}

function tracksByTimeRange(state = {}, action) {
  switch (action.type) {
    case RECEIVE_TRACKS:
    case REQUEST_TRACKS:
      return { ...state, [action.timeRange]: tracks(state[action.timeRange], action) };
    default:
      return state;
  }
}

function features(state = {}, action) {
  switch (action.type) {
    case REQUEST_FEATURES:
      return { ...state, isFetching: true };
    case RECEIVE_FEATURES:
      return { ...state, isFetching: false, items: action.features };
    default:
      return state;
  }
}

const rootReducer = combineReducers({
  user,
  timeRange,
  artistCount,
  trackCount,
  genreCount,
  artists,
  artistsByTimeRange,
  tracks,
  tracksByTimeRange,
  features,
});

export default rootReducer;
