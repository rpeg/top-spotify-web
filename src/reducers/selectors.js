export const selectHasClickedCreate = (state) => state.hasClickedCreate;
export const selectUser = (state) => state.user;
export const selectTimeRangeName = (state) => state.timeRangeName;
export const selectGenreCount = (state) => state.genreCount;
export const selectArtistCount = (state) => state.artistCount;
export const selectTrackCount = (state) => state.trackCount;
export const selectStatsOptions = (state) => state.statsOptions;
export const selectDisplayProfile = (state) => state.displayProfile;
export const selectOptimizeTracks = (state) => state.optimizeTracks;
export const selectArtists = (state) => state.artists;
export const selectArtistsByCurrentTimeRange = (state) => state.artistsByTimeRangeName[state.timeRangeName];
export const selectTracks = (state) => state.tracks;
export const selectTracksByCurrentTimeRange = (state) => state.tracksByTimeRangeName[state.timeRangeName];
export const selectFeatures = (state) => state.features;
export const selectFeaturesByCurrentTimeRange = (state) => state.featuresByTimeRangeName[state.timeRangeName];
