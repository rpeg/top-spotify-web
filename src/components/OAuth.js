import React, { useState, useEffect } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import PropTypes from 'prop-types';
import { Button } from 'react-bootstrap';

import { setUser } from '../actions/actions';
import { API_URL } from '../config';
import { selectUser } from '../reducers/selectors';

const buttonStyle = {
  right: '30px', top: '30px', marginBottom: '30px', position: 'absolute',
};

const OAuth = ({ socket }) => {
  const [disabled, setDisabled] = useState(false);
  const user = useSelector(selectUser);
  const dispatch = useDispatch();

  let popup;

  useEffect(() => {
    socket.on('spotifyUser', (spotifyUser) => {
      if (popup) { popup.close(); }

      dispatch(setUser(spotifyUser));
    });
  });

  const checkPopup = () => {
    const check = setInterval(() => {
      if (!popup || popup.closed || popup.closed === undefined) {
        clearInterval(check);
        setDisabled(false);
      }
    }, 1000);
  };

  const openPopup = () => {
    const width = 600;
    const height = 600;
    const left = window.innerWidth / 2 - width / 2;
    const top = window.innerHeight / 2 - height / 2;
    const url = `${API_URL}/auth/login?socketId=${socket.id}`;

    return window.open(
      url,
      '',
      `toolbar=no, location=no, directories=no, status=no, menubar=no, 
      scrollbars=no, resizable=no, copyhistory=no, width=${width}, 
      height=${height}, top=${top}, left=${left}`,
    );
  };

  const startAuth = (e) => {
    if (!disabled) {
      e.preventDefault();
      popup = openPopup();
      checkPopup();
      setDisabled(true);
    }
  };

  const logout = () => {
    dispatch(setUser({}));
  };

  return (
    <div>
      {user && user.id ? (
        <Button variant="outline-primary" style={buttonStyle} onClick={logout}>
          Logout
        </Button>
      ) : (
        <Button variant="outline-primary" style={buttonStyle} onClick={startAuth}>
          Login
        </Button>
      )}
    </div>
  );
};

OAuth.propTypes = {
  socket: PropTypes.object.isRequired,
};

export default OAuth;
