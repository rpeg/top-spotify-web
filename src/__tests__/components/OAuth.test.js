import React from 'react';
import { shallow } from 'enzyme';

import OAuth from '../../components/OAuth';
import * as selectors from '../../reducers/selectors';

jest.mock('react-redux', () => ({
  useSelector: jest.fn((fn) => fn()),
  useDispatch: () => jest.fn(),
}));

const setup = ({ user = {} }) => {
  jest.spyOn(selectors, 'selectUser').mockReturnValue(user);
};

const socket = { id: 'abc', on: () => {} };

describe('OAuth', () => {
  afterEach(() => {
    jest.clearAllMocks();
  });

  afterAll(() => {
    jest.restoreAllMocks();
  });

  it('should not log errors in console', () => {
    setup({});
    const wrapper = shallow(<OAuth socket={socket} />);
    const spy = jest.spyOn(global.console, 'error');
    expect(wrapper).not.toBeNull();
    expect(spy).not.toHaveBeenCalled();
  });

  it('renders correctly without user', () => {
    setup({});
    const wrapper = shallow(<OAuth socket={socket} />);
    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find('.button-login').exists()).toBe(true);
    expect(wrapper.find('.button-logout').exists()).toBe(false);
  });

  it('renders correctly with user', () => {
    setup({
      user: {
        id: 'abc',
      },
    });

    const wrapper = shallow(<OAuth socket={socket} />);
    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find('.button-login').exists()).toBe(false);
    expect(wrapper.find('.button-logout').exists()).toBe(true);
  });
});
