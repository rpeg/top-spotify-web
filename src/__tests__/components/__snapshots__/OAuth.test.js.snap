// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OAuth renders correctly with user 1`] = `
<div>
  <Button
    active={false}
    className="button-logout"
    disabled={false}
    onClick={[Function]}
    style={
      Object {
        "marginBottom": "30px",
        "position": "absolute",
        "right": "30px",
        "top": "30px",
      }
    }
    type="button"
    variant="primary"
  >
    Logout
  </Button>
</div>
`;

exports[`OAuth renders correctly without user 1`] = `
<div>
  <Button
    active={false}
    className="button-login"
    disabled={false}
    onClick={[Function]}
    style={
      Object {
        "marginBottom": "30px",
        "position": "absolute",
        "right": "30px",
        "top": "30px",
      }
    }
    type="button"
    variant="primary"
  >
    Login
  </Button>
</div>
`;
